### PR TITLE
fix: read rpc config when using fork cheatcodes

### LINKED ISSUE #9162: feat(`cheatcodes`): add `--fork-retry` + `--fork-retry-backoff` equivalent to `vm.create*Fork`
### Component

Forge

### Describe the feature you would like

Most of our integration tests are multichain, meaning we use `vm.createFork` to set up the forks. We can't use the `--fork-url` option because we want to ensure that the fork IDs have the same order, regardless of which chain the tests belong to.

For example, we always instantiate the forks in the following order:

- ID 0: Moonbeam fork
- ID 1: Base fork
- ID 2: Optimism fork

When running a `Base` integration test, the `setUp` function first fetches the Moonbeam governor and executes any live proposals (that can be cross-chain).

This is why we cannot use `--fork-url` with `forge test`.

Please allow the `--fork-retry` and `--fork-retry-backoff` flags to be used without requiring the `--fork-url`.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by anukul**
## Motivation

Closes #9162 - Set retry config (backoff, max retries) from rpc configuration when a fork is created using cheatcodes

## Solution

Idea - https://github.com/foundry-rs/foundry/issues/9162#issuecomment-2535586933

Implementation:

Use two structs `RpcEndpoint` and `ResolvedRpcEndpoint` that both contain the same data, with the exception that environment variables are resolved to their values in `ResolvedRpcEndpoint`.

Extract shared fields out into a common struct field `RpcEndpointConfig` that doesn't need environment variable resolution.

### CODE REVIEW

File: crates/cheatcodes/src/evm/fork.rs [hunk 1]
```diff
@@ -326,9 +326,15 @@ fn create_fork_request(
...
-    let url = ccx.state.config.rpc_url(url_or_alias)?;
+    let rpc_endpoint = ccx.state.config.rpc_endpoint(url_or_alias)?;
+    let url = rpc_endpoint.url()?;
     let mut evm_opts = ccx.state.config.evm_opts.clone();
     evm_opts.fork_block_number = block;
+    evm_opts.fork_retries = rpc_endpoint.config.retries;
+    evm_opts.fork_retry_backoff = rpc_endpoint.config.retry_backoff;
+    if let Some(Ok(auth)) = rpc_endpoint.auth {
+        evm_opts.fork_headers = Some(vec![format!("Authorization: {auth}")]);
+    }
```
Comment (mattsse) [inline crates/cheatcodes/src/evm/fork.rs:337]: cool

File: crates/cheatcodes/src/config.rs [hunk 2]
```diff
@@ -185,33 +185,33 @@ impl CheatsConfig {
     ///  - Returns an error if `url_or_alias` is a known alias but references an unresolved env var.
     ///  - Returns an error if `url_or_alias` is not an alias but does not start with a `http` or
     ///    `ws` `scheme` and is not a path to an existing file
-    pub fn rpc_url(&self, url_or_alias: &str) -> Result<String> {
-        match self.rpc_endpoints.get(url_or_alias) {
-            Some(Ok(url)) => Ok(url.clone()),
-            Some(Err(err)) => {
+    pub fn rpc_endpoint(&self, url_or_alias: &str) -> Result<ResolvedRpcEndpoint> {
+        if let Some(endpoint) = self.rpc_endpoints.get(url_or_alias) {
+            let mut endpoint = endpoint.clone();
```
Comment (mattsse) [inline crates/cheatcodes/src/config.rs:190]: do we need this clone here?

looks like we can `return endpoint.try_resolve()` instead if unresolved?

File: crates/config/src/endpoints.rs [hunk 3]
```diff
@@ -383,46 +390,82 @@ impl<'de> Deserialize<'de> for RpcEndpointConfig {
...
+            return self
+        }
+        let mut new_endpoint = self.clone();
+        if let Err(err) = self.endpoint {
+            new_endpoint.endpoint = err.try_resolve()
+        }
+        if let Some(Err(err)) = &new_endpoint.auth {
+            new_endpoint.auth = Some(err.try_resolve())
+        }
+        new_endpoint
```
Comment (mattsse) [inline crates/config/src/endpoints.rs:450]: this looks slightly weird

this clone shouldn't be necessary here

File: crates/cheatcodes/src/config.rs [hunk 4]
```diff
@@ -185,33 +185,33 @@ impl CheatsConfig {
...
+            if endpoint.is_unresolved() {
```
Comment (mattsse) [inline crates/cheatcodes/src/config.rs:191]: isn't this check redundant because this is also called in try_resolve?

**Review by mattsse — CHANGES_REQUESTED** [review]
this is great,

a few nits/questions

File: crates/cheatcodes/src/config.rs (hunk 4 above)
Comment (anukul) [reply crates/cheatcodes/src/config.rs]: thanks for catching that, pushed some fixes

**Comment by zerosnacks** [comment]
Thanks @anukul! This looks great

File: crates/cheatcodes/src/config.rs [hunk 5]
```diff
@@ -187,12 +187,8 @@ impl CheatsConfig {
     ///    `ws` `scheme` and is not a path to an existing file
     pub fn rpc_endpoint(&self, url_or_alias: &str) -> Result<ResolvedRpcEndpoint> {
         if let Some(endpoint) = self.rpc_endpoints.get(url_or_alias) {
-            let mut endpoint = endpoint.clone();
-            if endpoint.is_unresolved() {
-                // try resolve again, by checking if env vars are now set
-                endpoint = endpoint.try_resolve();
-            }
-            Ok(endpoint)
+            // try resolve again, by checking if env vars are now set
```
Comment (mattsse) [inline crates/cheatcodes/src/config.rs:190]:
```suggestion
```

**Review by mattsse — APPROVED** [review]

**Pull request merged**